pulp_ansible/app/tasks/collections.py

```diff
@@ -173,3 +106,111 @@ def import_collection(artifact_pk):
+        """
+        with ProgressBar(message="Parsing Collection Metadata") as pb:
+            async for metadata in self._fetch_collections():
+
+                for version in metadata["versions"]:
+
+                    url_path = "download/%s-%s-%s.tar.gz" % (
+                        metadata["namespace"],
+                        metadata["name"],
+                        version["version"],
+                    )
+                    url = urlunparse(urlparse(metadata["href"])._replace(path=url_path))
+
+                    collection = Collection(name=metadata["name"], namespace=metadata["namespace"])
+
+                    collection_version = CollectionVersion(
+                        version=version["version"], collection=collection
+                    )
+
+                    relative_path = "%s/%s/%s.tar.gz" % (
+                        metadata["namespace"],
+                        metadata["name"],
+                        version["version"],
+                    )
+                    d_artifact = DeclarativeArtifact(
+                        artifact=Artifact(),
+                        url=url,
+                        relative_path=relative_path,
+                        remote=self.remote,
+                        deferred_download=self.deferred_download,
+                    )
+                    d_content = DeclarativeContent(
+                        content=collection_version, d_artifacts=[d_artifact]
+                    )
+                    pb.increment()
+                    await self.put(d_content)
+
+    async def _fetch_collections(self):
+        async for metadata in self._fetch_galaxy_pages():
+            for result in metadata["results"]:
+
+                response = requests.get(result["versions_url"])
+
+                collection = {
+                    "href": result["href"],
+                    "name": result["name"],
+                    "namespace": result["namespace"]["name"],
+                    "versions": response.json()["results"],
+                }
+                yield collection
+
+    async def _fetch_galaxy_pages(self):
+        """
+        Fetch the collections in a remote repository.
+
+        Returns:
+            async generator: dicts that represent pages from galaxy api
+
+        """
+        page_count = 0
+        remote = self.remote
+
+        with ProgressBar(message="Parsing Pages from Galaxy Collections API") as progress_bar:
+            downloader = remote.get_downloader(url=get_page_url(remote.url))
+            metadata = parse_metadata(await downloader.run())
+
+            page_count = math.ceil(float(metadata["count"]) / float(PAGE_SIZE))
+            progress_bar.total = page_count
+            progress_bar.save()
+
+            yield metadata
+            progress_bar.increment()
+
+            # Concurrent downloads are limited by aiohttp...
+            not_done = set(
+                remote.get_downloader(url=get_page_url(remote.url, page)).run()
```

Comment:
You mentioned the too-many open files and I think it's this that is causing it. Yes while aiohttp is limiting the concurrency of co-routines that are scheduled I think each co-route created (scheduled or not) will allocate file descriptors. I believe this loop could create a large number of downloaders.

The Stages API has these restrictors built in for the downloaders it creates because we also don't want to hold a large number of them in memory. So you probably want to integrate this get_downloader into the loop below so that the loop itself delays the generation of additional downloaders until some number of existing ones are completed.

Does this make sense?

Reply:
It makes sense, but I am confused, why this piece of code works on roles but is not working here

Reply:
Is it that the actual count of downloaders is effectively different since it's driven by the the page counts reported by the servers? Maybe run a comparison on the number of downloaders generated? Also you can look at the file descriptor count if you also monitor the process with `lsof` to see when they are being allocated for sure. Just a few ideas.

Reply:
Now I got it, it was because I was using: https://galaxy-dev.ansible.com/api/v2/collections/
which has `"count": 11731`. When I changed to: https://galaxy.ansible.com/api/v2/collections/ which has `"count": 32`, the code finally worked!

Reply:
@fabricio-aguiar So did instantiating a lot of downloaders cause the too many files issue? I think that would be a good bug to file if so.

Reply:
I've checked the `remote.download_concurrency` as you suggested and it is 20 for both, roles and collections